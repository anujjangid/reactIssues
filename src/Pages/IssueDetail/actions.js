import {
  ISSUE_FETCHING,
  ISSUE_SUCCESS,
  ISSUE_ERROR,
  ISSUE_COMMENTS_FETCHING,
  ISSUE_COMMENTS_SUCCESS,
  ISSUE_COMMENTS_ERROR,
} from '../constant';
import { github } from '../../Util/api';

export function fetchSingleIssue(dispatch, issueId) {
  fetch(`${github.issueList}/${issueId}`)
    .then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .then(response =>
      dispatch({
        type: ISSUE_SUCCESS,
        payload: response,
      }),
    )
    .catch(err =>
      dispatch({
        type: ISSUE_ERROR,
        payload: err,
      }),
    );
  return {
    type: ISSUE_FETCHING,
    payload: {},
  };
}

export function fetchIssueComments(dispatch, issueId) {
  fetch(`${github.issueList}/${issueId}/comments`)
    .then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    })
    .then(response =>
      dispatch({
        type: ISSUE_COMMENTS_SUCCESS,
        payload: response,
      }),
    )
    .catch(err =>
      dispatch({
        type: ISSUE_COMMENTS_ERROR,
        payload: err,
      }),
    );
  return {
    type: ISSUE_COMMENTS_FETCHING,
    payload: {},
  };
}
