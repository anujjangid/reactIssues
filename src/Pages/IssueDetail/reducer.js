import {
  ISSUE_FETCHING,
  ISSUE_SUCCESS,
  ISSUE_ERROR,
  ISSUE_COMMENTS_FETCHING,
  ISSUE_COMMENTS_SUCCESS,
  ISSUE_COMMENTS_ERROR
} from "../constant";

const InitialState = {
  issues: {
    isLoading: false,
    data: [],
    comments: [],
    err: {}
  }
};

export default (state = InitialState, action) => {
  switch (action.type) {
    case ISSUE_FETCHING:
      return {
        ...state,
        isLoading: true,
        data: []
      };
    case ISSUE_SUCCESS:
      return {
        ...state,
        isLoading: false,
        data: action.payload
      };
    case ISSUE_ERROR:
      return {
        ...state,
        isLoading: false,
        err: action.payload
      };
    case ISSUE_COMMENTS_FETCHING:
      return {
        ...state,
        isLoading: true,
        comments: []
      };
    case ISSUE_COMMENTS_SUCCESS:
      return {
        ...state,
        comments: action.payload
      };
    case ISSUE_COMMENTS_ERROR:
      return {
        ...state,
        err: action.payload,
        comments: []
      };
    default:
      return {
        ...state
      };
  }
};
