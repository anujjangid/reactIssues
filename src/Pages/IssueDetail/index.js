// @flow

import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchSingleIssue, fetchIssueComments } from './actions';
import { Link } from 'react-router-dom';
import moment from 'moment';

import { IssueListDetail, Comment } from '../../Components/styleComponent';

const stateProps = state => ({
  issue: state.issueDetail,
});

type Props = {
  issue: Object,
  issueId: String,
  fetchSingleIssue: String => {},
  fetchIssueComments: String => {},
};

const dispatchProps = (dispatch: Function) => ({
  fetchSingleIssue: (issueId: string) =>
    dispatch(fetchSingleIssue(dispatch, issueId)),
  fetchIssueComments: (issueId: string) =>
    dispatch(fetchIssueComments(dispatch, issueId)),
});

class IssueDetail extends Component<Props> {
  static defaultProps = {
    issue: {},
    issueId: '',
    fetchSingleIssue: () => {},
    fetchIssueComments: () => {},
  };

  componentDidMount() {
    const { issueId, fetchSingleIssue, fetchIssueComments } = this.props;
    if (issueId) {
      fetchSingleIssue(issueId);
      fetchIssueComments(issueId);
    }
  }

  render() {
    const { data, comments } = this.props.issue;
    return (
      <div className="page">
        <Link to="./">{'Back'}</Link>
        {data && (
          <IssueListDetail>
            <h1>
              <span className="title">{data.title}</span>
              <span className="number">{data.number && `#${data.number}`}</span>
            </h1>
            <div className="headerBase">
              {data.user && <a href={data.user.html_url}>{data.user.login}</a>}
              {` opened this issue ${data.created_at &&
                moment(data.created_at)
                  .endOf('day')
                  .fromNow()} ${
                data.comments ? ` - ${data.comments} comments` : ''
              }`}
            </div>

            <div className="comments">
              {comments &&
                comments.map(comment => (
                  <Comment key={comment.id}>
                    <div className="header">{comment.user.login}</div>
                    <div className="body">{comment.body}</div>
                  </Comment>
                ))}
            </div>
          </IssueListDetail>
        )}
      </div>
    );
  }
}

export default connect(
  stateProps,
  dispatchProps,
)(IssueDetail);
