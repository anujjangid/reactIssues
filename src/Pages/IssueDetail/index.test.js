import React from 'react';
import renderer from 'react-test-renderer';
import IssueDetail from './index';

import { BrowserRouter } from 'react-router-dom';

import { Provider } from 'react-redux';
import configureStore from '../../store';
const store = configureStore();

describe('Issue Detail Page', () => {
  const issue = {
    data: {
      created_at: '',
      user: {},
    },
    comments: '',
  };
  it('renders correctly with data', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <IssueDetail issue={issue} />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
