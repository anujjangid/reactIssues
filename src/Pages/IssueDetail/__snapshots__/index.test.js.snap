// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue Detail Page renders correctly with data 1`] = `
<div
  className="page"
>
  <a
    href="/"
    onClick={[Function]}
  >
    Back
  </a>
</div>
`;
